shared/src/commonMain/kotlin/io/customer/shared/tracking/queue/QueueDispatcher.kt

```diff
@@ -0,0 +1,154 @@
+package io.customer.shared.tracking.queue
+
+import io.customer.shared.database.TrackingTask
+import io.customer.shared.database.TrackingTaskQueryHelper
+import io.customer.shared.sdk.config.BackgroundQueueConfig
+import io.customer.shared.tracking.api.*
+import io.customer.shared.tracking.api.model.*
+import io.customer.shared.tracking.constant.Priority
+import io.customer.shared.tracking.constant.QueueTaskStatus
+import io.customer.shared.util.DateTimeUtil
+import io.customer.shared.util.Logger
+import io.customer.shared.work.*
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.datetime.Instant
+import kotlin.time.DurationUnit
+
+/**
+ * Class responsible for running queue operations. This class only receives requests to validate the
+ * queue and is responsible for following:
+ *
+ * - whether to run the queue now or wait (except when specifically asked to run).
+ * - maintain lock to avoid sending same tasks multiple times.
+ * - validate and update task status related to queue operations.
+ * - schedule timer for next queue request.
+ */
+internal interface QueueDispatcher {
+    suspend fun checkForPendingTasks(source: QueueTriggerSource)
+    suspend fun sendAllPending()
+}
+
+internal class QueueDispatcherImpl(
+    private val logger: Logger,
+    private val dateTimeUtil: DateTimeUtil,
+    override val jobExecutor: JobExecutor,
+    private val backgroundQueueConfig: BackgroundQueueConfig,
+    private val trackingTaskQueryHelper: TrackingTaskQueryHelper,
+    private val queueRunner: QueueRunner,
+    private val queueTimer: QueueTimer,
+) : QueueDispatcher, JobDispatcher {
+    private val mutex = Mutex()
```

Comment:
Can we use `synchronized` instead of a mutex?

Reply:
I agree. I [added a comment](https://github.com/customerio/mobile-shared/pull/5/files#r1042219779) for this too. As long we are updating tasks status in DB with lock, I don't see any scenario where tasks can be repeated or missed. In fact, they can run in parallel as you suggested.

When I added this lock initially, I was not really sure where we'll end up in this PR. But now, I too feel it can be avoided. But I need some time to implement and test this. So we'll surely try removing this, but in a separate PR after running some tests on it.

Reply:
Yup agreed, locking the queue can be avoided.

> As long we are updating tasks status in DB with lock

We can probably make that method `synchronized` ?

Reply:
Yeah, I think Database lock should be enough. But I'll need time to try and test this out. So will work on this in a separate PR.

Also, [synchronized is unfortunately only available for JVM](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/synchronized.html), so we'll have to look for other locking solutions if we need locking the queue in KMM.